ambry-store/src/main/java/com.github.ambry.store/StatsManager.java

```diff
@@ -0,0 +1,202 @@
+        throw new IOException(
+            "Failed to rename " + tempFile.getAbsolutePath() + " to " + statsOutputFile.getAbsolutePath());
+      }
+    } else {
+      throw new IOException("Temporary file creation failed when publishing stats " + tempFile.getAbsolutePath());
+    }
+  }
+
+  /**
+   * Fetch and aggregate stats from a given {@link Store}
+   * @param aggregatedSnapshot the {@link StatsSnapshot} to hold the aggregated result
+   * @param partitionId specifies the {@link Store} to be fetched from
+   * @param unreachableStores a {@link List} containing partition Ids that were unable to successfully fetch from
+   */
+  void collectAndAggregate(StatsSnapshot aggregatedSnapshot, PartitionId partitionId, List<String> unreachableStores) {
+    Store store = storageManager.getStore(partitionId);
+    if (store == null) {
+      unreachableStores.add(partitionId.toString());
+    } else {
+      try {
+        long fetchAndAggregatePerStoreStartTimeMs = time.milliseconds();
+        StatsSnapshot statsSnapshot = store.getStoreStats().getStatsSnapshot();
+        aggregate(aggregatedSnapshot, statsSnapshot);
+        metrics.fetchAndAggregateTimePerStore.update(time.milliseconds() - fetchAndAggregatePerStoreStartTimeMs);
+      } catch (StoreException e) {
+        unreachableStores.add(partitionId.toString());
+      }
+    }
+  }
+
+  /**
+   * Performs recursive aggregation of two {@link StatsSnapshot} and stores the result in the first one.
+   * @param baseSnapshot one of the addends and where the result will be
+   * @param newSnapshot the other addend to be added into the first {@link StatsSnapshot}
+   */
+  private void aggregate(StatsSnapshot baseSnapshot, StatsSnapshot newSnapshot) {
+    baseSnapshot.setValue(baseSnapshot.getValue() + newSnapshot.getValue());
+    if (baseSnapshot.getSubtree() == null) {
+      baseSnapshot.setSubtree(newSnapshot.getSubtree());
+    } else if (newSnapshot.getSubtree() != null) {
+      for (Map.Entry<String, StatsSnapshot> entry : newSnapshot.getSubtree().entrySet()) {
+        if (!baseSnapshot.getSubtree().containsKey(entry.getKey())) {
+          baseSnapshot.getSubtree().put(entry.getKey(), new StatsSnapshot(0L, null));
+        }
+        aggregate(baseSnapshot.getSubtree().get(entry.getKey()), entry.getValue());
+      }
+    }
+  }
+
+  /**
+   * Runnable class that collects, aggregate and publish stats via methods in StatsManager.
+   */
+  private class StatsAggregator implements Runnable {
+    private volatile boolean cancelled = false;
+
+    @Override
+    public void run() {
+      long totalFetchAndAggregateStartTimeMs = time.milliseconds();
+      StatsSnapshot aggregatedSnapshot = new StatsSnapshot(0L, null);
+      List<String> unreachableStores = new ArrayList<>();
+      Iterator<PartitionId> iterator = initialPartitionIds.iterator();
+      while (!cancelled && iterator.hasNext()) {
+        PartitionId partitionId = iterator.next();
```

Comment:
Do you think checking if store exists should be moved here(instead of having it within collectAndAggregate?

```
 while (!cancelled && iterator.hasNext()) {
        PartitionId partitionId = iterator.next();
        Store store = storageManager.getStore(partitionId);
        if (store == null) {
            unreachableStores.add(partitionId.toString());
        } else {
            collectAndAggregate(aggregatedSnapshot, store);  
        }
 }
```